File: julia/NIfTIZarr.jl/test/test_pyramid_level.jl
using Test
using Random
using LinearAlgebra
import NIfTI
using Zarr
using NIfTIZarr

function count_levels(nifti_zarr::String)
    """
    In a .nii.zarr path, the number of folders should be the number of the levels plus 1(nifti folder)
    """
    for (root, dirs, files) in walkdir(nifti_zarr)
        return length(dirs) - 1
    end
end

temp_dir = mktempdir()
output_zarr = joinpath(temp_dir, "output.nii.zarr")
img_data = rand(Float64,(16, 32, 64))
ni = NIfTI.NIVolume(img_data)

@testset "No NB Level" begin
    nii2zarr(ni, output_zarr, nb_levels=1)
    @test count_levels(output_zarr) == 1
end

@testset "Constant NB Level" begin
    nii2zarr(ni, output_zarr, nb_levels=2)
    @test count_levels(output_zarr) == 2
end

@testset "Chunk Size 16" begin
    nii2zarr(ni, output_zarr, nb_levels=-1, chunk=16)
    @test count_levels(output_zarr) == 3
end

@testset "Chunk Size 24" begin
    nii2zarr(ni, output_zarr, nb_levels=-1, chunk=24)
    @test count_levels(output_zarr) == 3
end

@testset "Chunk Size 32" begin
    nii2zarr(ni, output_zarr, nb_levels=-1, chunk=32)
    @test count_levels(output_zarr) == 2
end

@testset "Chunk Size 48" begin
    nii2zarr(ni, output_zarr, nb_levels=-1, chunk=48)
    @test count_levels(output_zarr) == 2
end

@testset "Chunk Size 64" begin
    nii2zarr(ni, output_zarr, nb_levels=-1, chunk=64)
    @test count_levels(output_zarr) == 1
end

@testset "Chunk Size 96" begin
    nii2zarr(ni, output_zarr, nb_levels=-1, chunk=96)
    @test count_levels(output_zarr) == 1
end

@testset "Chunk Size 512" begin
    nii2zarr(ni, output_zarr, nb_levels=-1, chunk=512)
    @test count_levels(output_zarr) == 1
end

File: python/tests/test_pyramid_level.py
import os
import tempfile
import unittest

import numpy as np
from nibabel import Nifti1Image

from niizarr import nii2zarr


def count_levels(nifti_zarr):
    """
    In a .nii.zarr path, the number of folders should be the number of the levels plus 1(nifti folder)
    """
    for root, dir, file in os.walk(nifti_zarr):
        return len(dir) - 1


class TestPyramidLevel(unittest.TestCase):
    def setUp(self):
        self.temp_dir = tempfile.TemporaryDirectory()
        self.output_zarr = os.path.join(self.temp_dir.name, 'output.nii.zarr')
        img = np.random.rand(16, 32, 64)
        self.ni = Nifti1Image(img, np.eye(4))

    def tearDown(self):
        self.temp_dir.cleanup()

    def test_no_nb_level(self):
        nii2zarr(self.ni, self.output_zarr, nb_levels=1)
        self.assertEqual(1, count_levels(self.output_zarr))

    def test_constant_nb_level(self):
        nii2zarr(self.ni, self.output_zarr, nb_levels=2)
        self.assertEqual(2, count_levels(self.output_zarr))

    def test_chunk_size_16(self):
        nii2zarr(self.ni, self.output_zarr, nb_levels=-1, chunk=16)
        self.assertEqual(3, count_levels(self.output_zarr))

    def test_chunk_size_24(self):
        nii2zarr(self.ni, self.output_zarr, nb_levels=-1, chunk=24)
        self.assertEqual(3, count_levels(self.output_zarr))

    def test_chunk_size_32(self):
        nii2zarr(self.ni, self.output_zarr, nb_levels=-1, chunk=32)
        self.assertEqual(2, count_levels(self.output_zarr))

    def test_chunk_size_48(self):
        # 64 -> 32 OK
        nii2zarr(self.ni, self.output_zarr, nb_levels=-1, chunk=48)
        self.assertEqual(2, count_levels(self.output_zarr))

    def test_chunk_size_64(self):
        nii2zarr(self.ni, self.output_zarr, nb_levels=-1, chunk=64)
        self.assertEqual(1, count_levels(self.output_zarr))

    def test_chunk_size_96(self):
        nii2zarr(self.ni, self.output_zarr, nb_levels=-1, chunk=96)
        self.assertEqual(1, count_levels(self.output_zarr))

    def test_chunk_size_512(self):
        nii2zarr(self.ni, self.output_zarr, nb_levels=-1, chunk=512)
        self.assertEqual(1, count_levels(self.output_zarr))


if __name__ == '__main__':
    unittest.main()
